Assertion failed 'farthestRefPhysRegRecord != nullptr' /dotnet/coreclr/src/jit/lsra.cpp:3637

I'm using the x86 version of the CoreCLR to run CoreRT and compile tests for ARM.

After the commit:
> b39a5b2cfad4620ae23cda84186a36ffb463cd5a 

compilation fails (for all tests) with message

> cd /overlay.arm.Debug && ./corerun ilc.dll --codegenopt "AltJitNgen=*" @/overlay.arm.Debug/pow3/base.ilc.rsp
/dotnet/coreclr/src/jit/lsra.cpp:3637
Assertion failed 'farthestRefPhysRegRecord != nullptr' in 'Sys:DoubleToString(double,int,int,int):int' (IL size 14)

The folder /overlay.arm.Debug/ contains necessary parts from CoreRT, CoreCLR and CoreFX to 
cross-compile IL binary to native ARM code.  I've followed the following guide: [https://github.com/dotnet/corert/blob/master/Documentation/cross-building.md](url)
The `@/overlay.arm.Debug/pow3/base.ilc.rsp` is dependency file, generated by the `msbuild`. 
The pow3 test is one of the CoreCRL's tests for CoreRT. The tests are downloaded by `/dotnet/corert/tests/runtest.sh`  script and could be run by the following command `./tests/runtest.sh -coreclr all`

I've checked with the following versions of the CoreCRL, CoreRT and CoreFX:

CoreCLR: 6bf04a47badd74646e21e70f4e9267c71b7bfd08
CoreFX: ccdb5a48a49d3e932aa00d02749777464f4c8d5b
CoreRT: a028615abe80057e413d632a7c9257208b7c6865

I'm attempting to build using the instructions here: https://github.com/dotnet/corert/blob/master/Documentation/cross-building.md but for arm instead of armel. I'm getting the following error from `./build.sh arm debug verbose clean cross`:
```
Unhandled Exception: System.DllNotFoundException: Unable to load shared library 'jitinterface' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libjitinterface: cannot open shared object file: No such file or directory
     at Internal.JitInterface.CorInfoImpl.GetJitHost(IntPtr configProvider)
     at Internal.JitInterface.CorInfoImpl..ctor(Compilation compilation, JitConfigProvider jitConfig) in /git/corert/src/JitInterface/src/CorInfoImpl.cs:line 113
     at ILCompiler.RyuJitCompilation.CompileInternal(String outputFile, ObjectDumper dumper) in /git/corert/src/ILCompiler.Compiler/src/Compiler/RyuJitCompilation.cs:line 37
     at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in /git/corert/src/ILCompiler.Compiler/src/Compiler/Compilation.cs:line 329
     at ILCompiler.Program.Run(String[] args) in /git/corert/src/ILCompiler/src/Program.cs:line 493
     at ILCompiler.Program.Main(String[] args) in /git/corert/src/ILCompiler/src/Program.cs:line 649
  Aborted (core dumped)
/git/corert/bin/Linux.arm.Debug/build/Microsoft.NETCore.Native.targets(183,5): error MSB3073: The command ""/git/corert/tests/../bin/Linux.arm.Debug/tools/ilc" @"/git/corert/tests/src/Simple/Threading/obj/Debug/arm/native/Threading.ilc.rsp"" exited with code 134. [/git/corert/tests/src/Simple/Threading/Threading.csproj]
```

I ran into repeated issues in attempting to build arm.
I got farther in attempting to build armel, but now the corert build step "./build.sh armel debug verbose cross" immediately fails with:
```

Setting up directories for build
Checking pre-requisites...
Commencing build of corert native components for Linux.armel.Debug
/git/corert/bin/obj/Native/Linux.armel.Debug /git/corert
Invoking cmake with arguments: "/git/corert/buildscripts/.." Debug
loading initial cache file /git/corert/buildscripts/../cross/tryrun.cmake
Re-run cmake no build system arguments
-- The C compiler identification is Clang 3.9.1
-- The CXX compiler identification is Clang 3.9.1
-- Check for working C compiler: /usr/bin/clang-3.9
-- Check for working C compiler: /usr/bin/clang-3.9 -- broken
CMake Error at /usr/share/cmake-3.5/Modules/CMakeTestCCompiler.cmake:61 (message):
  The C compiler "/usr/bin/clang-3.9" is not able to compile a simple test
  program.

  It fails with the following output:

   Change Dir: /git/corert/bin/obj/Native/Linux.armel.Debug/CMakeFiles/CMakeTmp



  Run Build Command:"/usr/bin/make" "cmTC_028ba/fast"

  /usr/bin/make -f CMakeFiles/cmTC_028ba.dir/build.make
  CMakeFiles/cmTC_028ba.dir/build

  make[1]: Entering directory
  '/git/corert/bin/obj/Native/Linux.armel.Debug/CMakeFiles/CMakeTmp'

  Building C object CMakeFiles/cmTC_028ba.dir/testCCompiler.c.o

  /usr/bin/clang-3.9 --target=arm-linux-gnueabi -isystem
  /git/corert/buildscripts/../cross/rootfs/armel/usr/lib/gcc/armv7l-tizen-linux-gnueabi/6.2.1/include/c++
  -isystem
  /git/corert/buildscripts/../cross/rootfs/armel/usr/lib/gcc/armv7l-tizen-linux-gnueabi/6.2.1/include/c++/armv7l-tizen-linux-gnueabi
  --sysroot=/git/corert/buildscripts/../cross/rootfs/armel
  --target=arm-linux-gnueabi
  --gcc-toolchain=/git/corert/buildscripts/../cross/rootfs/armel/usr -mthumb
  -mfpu=vfpv3 -mfloat-abi=softfp -Wno-deprecated-declarations
  -D__extern_always_inline=inline -o
  CMakeFiles/cmTC_028ba.dir/testCCompiler.c.o -c
  /git/corert/bin/obj/Native/Linux.armel.Debug/CMakeFiles/CMakeTmp/testCCompiler.c


  Linking C executable cmTC_028ba

  /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_028ba.dir/link.txt
  --verbose=1

  /usr/bin/clang-3.9 --target=arm-linux-gnueabi
  --sysroot=/git/corert/buildscripts/../cross/rootfs/armel
  --gcc-toolchain=/git/corert/buildscripts/../cross/rootfs/armel/usr
  --target=arm-linux-gnueabi -fuse-ld=gold
  -B/git/corert/buildscripts/../cross/rootfs/armel/usr/lib/gcc/armv7l-tizen-linux-gnueabi/6.2.1
  -L/git/corert/buildscripts/../cross/rootfs/armel/lib
  -L/git/corert/buildscripts/../cross/rootfs/armel/usr/lib
  -L/git/corert/buildscripts/../cross/rootfs/armel/usr/lib/gcc/armv7l-tizen-linux-gnueabi/6.2.1
  CMakeFiles/cmTC_028ba.dir/testCCompiler.c.o -o cmTC_028ba -rdynamic

  /usr/bin/ld.gold: fatal error:
  /git/corert/buildscripts/../cross/rootfs/armel/usr/lib/../lib/crt1.o:
  unsupported ELF machine number 40

  clang: error: linker command failed with exit code 1 (use -v to see
  invocation)

  CMakeFiles/cmTC_028ba.dir/build.make:97: recipe for target 'cmTC_028ba'
  failed

  make[1]: *** [cmTC_028ba] Error 1

  make[1]: Leaving directory
  '/git/corert/bin/obj/Native/Linux.armel.Debug/CMakeFiles/CMakeTmp'

  Makefile:126: recipe for target 'cmTC_028ba/fast' failed

  make: *** [cmTC_028ba/fast] Error 2

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:2 (project)
```
Any suggestions?
cc @dotnet/corert-contrib

--- a/src/jit/gentree.h
+++ b/src/jit/gentree.h
@@ -6030,16 +6030,18 @@ inline regNumber GenTree::GetRegByIndex(int regIndex)
         return AsCall()->GetRegNumByIdx(regIndex);
     }
 
+#if defined(_TARGET_ARMARCH_)
+    if (OperIsPutArgSplit())
+    {
+        return AsPutArgSplit()->GetRegNumByIdx(regIndex);
+    }
+#endif
 #if defined(_TARGET_ARM_)
     if (OperIsMultiRegOp())
     {
         return AsMultiRegOp()->GetRegNumByIdx(regIndex);
     }
-    else if (OperIsPutArgSplit())
-    {
-        return AsPutArgSplit()->GetRegNumByIdx(regIndex);
-    }
-    else if (OperIs(GT_COPY, GT_RELOAD))
+    if (OperIs(GT_COPY, GT_RELOAD))
     {
         return AsCopyOrReload()->GetRegNumByIdx(regIndex);
     }

--- a/src/jit/gentree.h
+++ b/src/jit/gentree.h
@@ -1721,6 +1721,9 @@ public:
     // Returns true if it is a node returning its value in more than one register
     inline bool IsMultiRegNode() const;
 
+    // Returns the regIndex'th register defined by a possibly-multireg node.
+    regNumber GetRegByIndex(int regIndex);
+
     // Returns true if it is a GT_COPY or GT_RELOAD node
     inline bool IsCopyOrReload() const;
 
@@ -6004,6 +6007,45 @@ inline bool GenTree::IsMultiRegNode() const
 #endif
 
     return false;
+}
+
+//-----------------------------------------------------------------------------------
+// GetRegByIndex: Get a specific register, based on regIndex, that is produced
+//                by this node.
+//
+// Arguments:
+//     regIndex - which register to return (must be 0 for non-multireg nodes)
+//
+// Return Value:
+//     The register, if any, assigned to this index for this node.
+//
+inline regNumber GenTree::GetRegByIndex(int regIndex)
+{
+    if (regIndex == 0)
+    {
+        return gtRegNum;
+    }
+    if (IsMultiRegCall())
+    {
+        return AsCall()->GetRegNumByIdx(regIndex);
+    }
+
+#if defined(_TARGET_ARM_)
+    if (OperIsMultiRegOp())
+    {
+        return AsMultiRegOp()->GetRegNumByIdx(regIndex);
+    }
+    else if (OperIsPutArgSplit())
+    {
+        return AsPutArgSplit()->GetRegNumByIdx(regIndex);
+    }
+    else if (OperIs(GT_COPY, GT_RELOAD))
+    {
+        return AsCopyOrReload()->GetRegNumByIdx(regIndex);
+    }
+#endif
+    assert(!"Invalid regIndex for GetRegFromMultiRegNode");
+    return REG_NA;
 }
 
 //-------------------------------------------------------------------------

--- a/src/jit/gentree.h
+++ b/src/jit/gentree.h
@@ -6030,7 +6030,7 @@ inline regNumber GenTree::GetRegByIndex(int regIndex)
         return AsCall()->GetRegNumByIdx(regIndex);
     }
 
-#if defined(_TARGET_ARMARCH_)
+#if FEATURE_ARG_SPLIT
     if (OperIsPutArgSplit())
     {
         return AsPutArgSplit()->GetRegNumByIdx(regIndex);

--- a/src/jit/lower.cpp
+++ b/src/jit/lower.cpp
@@ -1055,13 +1055,17 @@ GenTree* Lowering::NewPutArg(GenTreeCall* call, GenTree* arg, fgArgTabEntry* inf
         putArg = new (comp, GT_PUTARG_SPLIT)
             GenTreePutArgSplit(arg, info->slotNum PUT_STRUCT_ARG_STK_ONLY_ARG(info->numSlots), info->numRegs,
                                call->IsFastTailCall(), call);
-        putArg->gtRegNum = info->regNum;
 
         // If struct argument is morphed to GT_FIELD_LIST node(s),
         // we can know GC info by type of each GT_FIELD_LIST node.
         // So we skip setting GC Pointer info.
         //
         GenTreePutArgSplit* argSplit = putArg->AsPutArgSplit();
+        for (unsigned regIndex = 0; regIndex < info->numRegs; regIndex++)
+        {
+            argSplit->SetRegNumByIdx(info->getRegNum(regIndex), regIndex);
+        }
+
         if (arg->OperGet() == GT_OBJ)
         {
             BYTE*       gcLayout = nullptr;

--- a/src/jit/lsra.cpp
+++ b/src/jit/lsra.cpp
@@ -9109,7 +9109,7 @@ void LinearScan::TupleStyleDump(LsraTupleDumpMode mode)
                                 printf("\n                               Use:");
                                 interval->microDump();
                                 printf("(#%d)", currentRefPosition->rpNum);
-                                if (currentRefPosition->isFixedRegRef)
+                                if (currentRefPosition->isFixedRegRef && !interval->isInternal)
                                 {
                                     assert(genMaxOneBit(currentRefPosition->registerAssignment));
                                     assert(lastFixedRegRefPos != nullptr);

--- a/src/jit/lsraarmarch.cpp
+++ b/src/jit/lsraarmarch.cpp
@@ -283,7 +283,7 @@ int LinearScan::BuildCall(GenTreeCall* call)
             assert(regCount == curArgTabEntry->numRegs);
             for (unsigned int i = 0; i < regCount; i++)
             {
-                BuildUse(argNode, genRegMask(argNode->gtRegNum), i);
+                BuildUse(argNode, genRegMask(argNode->AsPutArgSplit()->GetRegNumByIdx(i)), i);
             }
             srcCount += regCount;
         }

--- a/src/jit/lsrabuild.cpp
+++ b/src/jit/lsrabuild.cpp
@@ -401,7 +401,7 @@ void LinearScan::applyCalleeSaveHeuristics(RefPosition* rp)
 //    incorrectly allocate that register.
 //    TODO-CQ: This means that we may often require a copy at the use of this node's result.
 //    This case could be moved to BuildRefPositionsForNode, at the point where the def RefPosition is
-//    created, causing a RefTypeFixedRef to be added at that location. This, however, results in
+//    created, causing a RefTypeFixedReg to be added at that location. This, however, results in
 //    more PhysReg RefPositions (a throughput impact), and a large number of diffs that require
 //    further analysis to determine benefit.
 //    See Issue #11274.
@@ -516,12 +516,16 @@ RefPosition* LinearScan::newRefPosition(
 {
     RefPosition* newRP = newRefPositionRaw(theLocation, theTreeNode, theRefType);
 
-    newRP->setReg(getRegisterRecord(reg));
+    RegRecord* regRecord = getRegisterRecord(reg);
+    newRP->setReg(regRecord);
     newRP->registerAssignment = mask;
 
     newRP->setMultiRegIdx(0);
     newRP->setAllocateIfProfitable(false);
 
+    // We can't have two RefPositions on a RegRecord at the same location, unless they are different types.
+    assert((regRecord->lastRefPosition == nullptr) || (regRecord->lastRefPosition->nodeLocation < theLocation) ||
+           (regRecord->lastRefPosition->refType != theRefType));
     associateRefPosWithInterval(newRP);
 
     DBEXEC(VERBOSE, newRP->dump());
@@ -579,8 +583,9 @@ RefPosition* LinearScan::newRefPosition(Interval*    theInterval,
     bool insertFixedRef  = false;
     if (isFixedRegister)
     {
-        // Insert a RefTypeFixedReg for any normal def or use (not ParamDef or BB)
-        if (theRefType == RefTypeUse || theRefType == RefTypeDef)
+        // Insert a RefTypeFixedReg for any normal def or use (not ParamDef or BB),
+        // but not an internal use (it will already have a FixedRef for the def).
+        if ((theRefType == RefTypeDef) || ((theRefType == RefTypeUse) && !theInterval->isInternal))
         {
             insertFixedRef = true;
         }
@@ -2322,6 +2327,12 @@ RefPosition* LinearScan::BuildDef(GenTree* tree, regMaskTP dstCandidates, int mu
 {
     assert(!tree->isContained());
     RegisterType type = getDefType(tree);
+
+    if (dstCandidates != RBM_NONE)
+    {
+        assert((tree->gtRegNum == REG_NA) || (dstCandidates == genRegMask(tree->GetRegByIndex(multiRegIdx))));
+    }
+
 #ifdef FEATURE_MULTIREG_ARGS_OR_RET
     if (tree->TypeGet() == TYP_STRUCT)
     {
@@ -2501,10 +2512,6 @@ RefPosition* LinearScan::BuildUse(GenTree* operand, regMaskTP candidates, int mu
     Interval* interval;
     bool      regOptional = operand->IsRegOptional();
 
-    if (operand->gtRegNum != REG_NA)
-    {
-        candidates = genRegMask(operand->gtRegNum);
-    }
     if (isCandidateLocalRef(operand))
     {
         interval = getIntervalForLocalVarNode(operand->AsLclVarCommon());

